### PR TITLE
48 Add support for 'Bits', like we do 'Sats'

### PR DESCRIPTION
**Pull request by paOol**
https://github.com/PayButton/paybutton/issues/98

&

https://github.com/PayButton/paybutton/issues/48

for this PR.

### CODE REVIEW

**Review by Klakurka — CHANGES_REQUESTED** [review]
- Bits support 2 decimals, so 2 decimals should be showing. Right now 0 decimals show.
- If a number >99999999 is entered, the amount gets treated as satoshis... it should just be bch, satoshis, bits, etc all the time depending on the currency selected. We can maybe set a limit on the number at some point. I'm not sure what's best for that.

File: react/src/util/satoshis.ts [hunk 1]
```diff
@@ -1,10 +1,104 @@
...
+  let BCHval = '';
+  let satoshiVal = 0;
+  const bchObj = new BigNumber(amount).decimalPlaces(BCHdecimals);
+  const inBCH = new BigNumber(amount)
+    .dividedBy(10 ** BCHdecimals)
+    .toFixed(BCHdecimals);
+  const satoshis = new BigNumber(`${amount}e+${BCHdecimals}`);
+
+  const { c } = bchObj;
+  // coefficient
```
[the remark this answers is not in the document]
Comment (paOol) [reply react/src/util/satoshis.ts:39]: it's to get this object 
image
versus doing a bunch of formatting in the react side.

File: react/src/components/Widget/WidgetContainer.tsx [hunk 2]
```diff
@@ -166,56 +175,35 @@ export const WidgetContainer: React.FC<WidgetContainerProps> = withSnackbar(
...
       if (!active) return;
 
-      if (!props.amount || ['BCH', 'SAT'].includes(currency)) {
+      if (!props.amount || ['BCH', 'SAT', 'bits'].includes(currency)) {
         setLoading(false);
-        return;
       }
 
-      setLoading(true);
+      // setLoading(true);
```
[the remark this answers is not in the document]
Comment (paOol) [reply react/src/components/Widget/WidgetContainer.tsx]: removed

**Comment by Klakurka** [comment]
- Wrong math: 
image
- The loading spinner is gone in this branch for some reason. It's supposed to have a spinner in the middle with a white square in place of the QR code, with 'Loading...' in the title section. In this branch, it shows a QR code for an amount of 0 briefly, as well as shows `Send $X.00 USD = BCH` which is not great.
- The title section where the price conversion is displayed is getting a bit long with the extra symbols + decimals, but that can be addressed with #80.

**Comment by Klakurka** [comment]
- For some reason the loading finishes, then it sits on the "0" QR code and title bar text for about half a second before switching to the proper QR code and title text. It happens every time, reliably. Must have `amount` filled in and `currency` set to fiat.

**Comment by Klakurka** [comment]
It's still kinda happening, just not every time (maybe 25% of the time).

It's definitely better but still probably should be fixed.

**Comment by Klakurka** [comment]
Still happening for me.

I'm testing using Storybook with `Currency: CAD` and `Amount: 55`.

**Comment by paOol** [comment]
After further investigation,
the problem happens because multiple network calls are made to the rest price endpoint.
This is unavoidable in the current <Widget> <WidgetContainer> setup.
The reason it only happens some times is because the rest API will sometimes return a different price value in those multiple network calls. 

One easy solution would be to switch the endpoint to a different api that does not update as quickly (even if the endpoint is cached by 5 seconds this problem wouldn't exist).

The other is lengthy and would require refactoring the Widget component entirely.

I think if the other changes look good, they should be merged in and the price endpoint should be made into its own issue.

**Pull request merged**